src/undocommand.js

```diff
@@ -0,0 +1,200 @@
+/**
+ * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+import Command from '../command/command.js';
+
+/**
+ * Undo command stores batches in itself and is able to and apply reverted versions of them on the document.
+ *
+ * undo.UndoCommand
+ */
+export default class UndoCommand extends Command {
+	/**
+	 * @see engine.command.Command
+	 * @param {engine.Editor} editor
+	 */
+	constructor( editor ) {
+		super( editor );
+
+		/**
+		 * Batches which are saved by the command. They can be reversed.
+		 *
+		 * @private
+		 * @member {Array.<engine.treeModel.Batch>} undo.UndoCommand#_batchStack
+		 */
+		this._batchStack = [];
+
+		/**
+		 * For each batch, it stores the information about selection needed to recreate it after undo.
+		 *
+		 * @private
+		 * @member {WeakMap} undo.UndoCommand#_batchSelection
+		 */
+		this._batchSelection = new WeakMap();
```

Comment:
This implementation went through a few iterations and at one point I had to have it in `WeakMap` cause the selection for given batch might get updated. At this stage, I don't need to recognize selection state by `Batch` instance, so as you said, it can be merged with `_batchStack`.